### PR TITLE
Add support for building forge jars

### PR DESCRIPTION
**Pull request by babbaj**
a bit scuffed just need to run build with `-Pbaritone.forge_build`

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -83,7 +86,13 @@ task sourceJar(type: Jar, dependsOn: classes) {
 
 minecraft {
     mappings channel: 'snapshot', version: '20200213-1.15.1'
-    reobfMappings 'notch'
+
+    if (getProject().hasProperty("baritone.forge_build")) {
+        println 'searge mappings!!'
```
Comment (leijurv) [inline build.gradle]: lol

File: build.gradle [hunk 2]
```diff
@@ -178,14 +188,48 @@ jar {
...
+    rename {'client-srg.jar'}
+}
+
+task createSrgMc(type: RenameJarInPlace) {
+    setInput(new File(copyMcJar.getOutputs().getFiles().getSingleFile(), "client-srg.jar"))
+    setClasspath(files({getClientJar()}))
+
+    // fork
+    setMappingType(net.minecraftforge.gradle.common.util.MappingFile.Mapping.SEARGE)
+    setJarTask('nword')
```
Comment (leijurv) [inline build.gradle]: no

File: buildSrc/src/main/java/baritone/gradle/task/CreateDistTask.java [hunk 3]
```diff
@@ -53,12 +53,15 @@ protected void exec() throws Exception {
...
-        Files.copy(this.artifactApiPath,             api,             REPLACE_EXISTING);
-        Files.copy(this.artifactStandalonePath,      standalone,      REPLACE_EXISTING);
-        Files.copy(this.artifactUnoptimizedPath,     unoptimized,     REPLACE_EXISTING);
+        // TODO: dont copy files that dont exist
+        Files.copy(this.artifactApiPath,         api,         REPLACE_EXISTING);
+        Files.copy(this.artifactStandalonePath,  standalone,  REPLACE_EXISTING);
+        Files.copy(this.artifactUnoptimizedPath, unoptimized, REPLACE_EXISTING);
 
         // Calculate all checksums and format them like "shasum"
+        // TODO: get the hash for both forge and non forge jars
```
Comment (leijurv) [inline buildSrc/src/main/java/baritone/gradle/task/CreateDistTask.java]: cring, necessary for paranoia gpg

File: src/api/java/baritone/api/BaritoneAPI.java [hunk 4]
```diff
@@ -37,9 +34,11 @@
         settings = new Settings();
         SettingsUtil.readAndApply(settings);
 
-        ServiceLoader<IBaritoneProvider> baritoneLoader = ServiceLoader.load(IBaritoneProvider.class);
-        Iterator<IBaritoneProvider> instances = baritoneLoader.iterator();
-        provider = instances.next();
+        try {
```
Comment (leijurv) [inline src/api/java/baritone/api/BaritoneAPI.java]: extremely cringe

File: src/launch/java/baritone/launch/mixins/MixinLivingEntity.java [hunk 5]
```diff
@@ -32,7 +32,8 @@
 import org.spongepowered.asm.mixin.injection.Redirect;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
 
-import static org.spongepowered.asm.lib.Opcodes.GETFIELD;
+import static org.objectweb.asm.Opcodes.GETFIELD;
+//import static org.spongepowered.asm.lib.Opcodes.GETFIELD;
```
Comment (leijurv) [inline src/launch/java/baritone/launch/mixins/MixinLivingEntity.java]: remove these comments

File: src/launch/resources/META-INF/mods.toml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+# The name of the mod loader type to load - for regular FML @Mod mods it should be javafml
+modLoader="javafml" #mandatory
+# A version range to match for said mod loader - for regular FML @Mod it will be the forge version
+loaderVersion="[31,)" #mandatory This is typically bumped every Minecraft version by Forge. See our download page for lists of versions.
+# A URL to refer people to when problems occur with this mod
+issueTrackerURL="http://my.issue.tracker/" #optional
+# A list of mods - how many allowed here is determined by the individual mod loader
+[[mods]] #mandatory
+# The modid of the mod
+modId="baritbone" #mandatory
```
Comment (leijurv) [inline src/launch/resources/META-INF/mods.toml]: baritoe

File: src/launch/resources/META-INF/mods.toml [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+# There are several mandatory fields (#mandatory), and many more that are optional (#optional).
+# The overall format is standard TOML format, v0.5.0.
+# Note that there are a couple of TOML lists in this file.
+# Find more information on toml format here:  https://github.com/toml-lang/toml
+# The name of the mod loader type to load - for regular FML @Mod mods it should be javafml
+modLoader="javafml" #mandatory
+# A version range to match for said mod loader - for regular FML @Mod it will be the forge version
+loaderVersion="[31,)" #mandatory This is typically bumped every Minecraft version by Forge. See our download page for lists of versions.
+# A URL to refer people to when problems occur with this mod
+issueTrackerURL="http://my.issue.tracker/" #optional
```
Comment (leijurv) [inline src/launch/resources/META-INF/mods.toml]: put dis in

File: src/launch/resources/META-INF/mods.toml [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+# A text field displayed in the mod UI
+authors="Love, Cheese and small house plants" #optional
+# The description text for the mod (multi line!) (#mandatory)
+description='''
+This is a long form description of the mod. You can write whatever you want here
+
+Have some lorem ipsum.
+
+Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed mollis lacinia magna. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Sed sagittis luctus odio eu tempus. Interdum et malesuada fames ac ante ipsum primis in faucibus. Pellentesque volutpat ligula eget lacus auctor sagittis. In hac habitasse platea dictumst. Nunc gravida elit vitae sem vehicula efficitur. Donec mattis ipsum et arcu lobortis, eleifend sagittis sem rutrum. Cras pharetra quam eget posuere fermentum. Sed id tincidunt justo. Lorem ipsum dolor sit amet, consectetur adipiscing elit.
+'''
```
Comment (leijurv) [inline src/launch/resources/META-INF/mods.toml]: fix all dis

File: src/api/java/baritone/api/BaritoneAPI.java (hunk 4 above)
Comment (babbaj) [reply src/api/java/baritone/api/BaritoneAPI.java]: > .class is CRING

proguard is CRING

**Comment by babbaj** [comment]
> .class is CRING

proguard is CRING

File: build.gradle [hunk 9]
```diff
@@ -178,14 +186,47 @@ jar {
...
+    rename {'client-srg.jar'}
+}
+
+task createSrgMc(type: RenameJarInPlace) {
+    setInput(new File(copyMcJar.getOutputs().getFiles().getSingleFile(), "client-srg.jar"))
+    setClasspath(files({getClientJar()}))
+
+    // fork
+    setMappingType(net.minecraftforge.gradle.common.util.MappingFile.Mapping.SEARGE)
+    setJarTask('trans alaskan pipeline')
```
Comment (leijurv) [inline build.gradle]: you cannot sneak the `n` in! it is just `trans alaska pipeline`

File: buildSrc/src/main/java/baritone/gradle/task/BaritoneGradleTask.java [hunk 10]
```diff
@@ -43,28 +44,43 @@
             ARTIFACT_STANDARD         = "%s-%s.jar",
             ARTIFACT_UNOPTIMIZED      = "%s-unoptimized-%s.jar",
             ARTIFACT_API              = "%s-api-%s.jar",
-            ARTIFACT_STANDALONE       = "%s-standalone-%s.jar";
+            ARTIFACT_STANDALONE       = "%s-standalone-%s.jar",
+            ARTIFACT_FORGE_UNOPTIMIZED = "%s-unoptimized-forge-%s.jar",
```
Comment (leijurv) [inline buildSrc/src/main/java/baritone/gradle/task/BaritoneGradleTask.java]: staggering of the = signs is a bad meme, either do it right or dont do it at all tbh...

File: buildSrc/src/main/java/baritone/gradle/task/CreateDistTask.java [hunk 11]
```diff
@@ -53,12 +53,14 @@ protected void exec() throws Exception {
...
-        Files.copy(this.artifactUnoptimizedPath,     unoptimized,     REPLACE_EXISTING);
+        // TODO: dont copy files that dont exist
+        Files.copy(this.artifactApiPath,         api,         REPLACE_EXISTING);
+        Files.copy(this.artifactStandalonePath,  standalone,  REPLACE_EXISTING);
+        Files.copy(this.artifactUnoptimizedPath, unoptimized, REPLACE_EXISTING);
 
         // Calculate all checksums and format them like "shasum"
-        List<String> shasum = Stream.of(api, standalone, unoptimized)
+        List<String> shasum = Files.walk(getRelativeFile("dist"), 1)
+                .filter(p -> p.toString().endsWith(".jar"))
```
Comment (leijurv) [inline buildSrc/src/main/java/baritone/gradle/task/CreateDistTask.java]: not good. my `dist/` folder contains a TON of old builds, but I only want it to print out the jars that it just wrote there this time

File: src/launch/java/baritone/launch/BaritoneMixinConnector.java [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Baritone.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package baritone.launch;
+
+import org.spongepowered.asm.mixin.Mixins;
+import org.spongepowered.asm.mixin.connect.IMixinConnector;
+
+
```
Comment (ZeroMemes) [inline src/launch/java/baritone/launch/BaritoneMixinConnector.java]: Fix double space

File: src/launch/resources/META-INF/mods.toml [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+# The name of the mod loader type to load - for regular FML @Mod mods it should be javafml
+modLoader="javafml" #mandatory
+# A version range to match for said mod loader - for regular FML @Mod it will be the forge version
+loaderVersion="[31,)" #mandatory This is typically bumped every Minecraft version by Forge. See our download page for lists of versions.
+# A URL to refer people to when problems occur with this mod
+issueTrackerURL="https://github.com/cabaletta/baritone/issues" #optional
+# A list of mods - how many allowed here is determined by the individual mod loader
+[[mods]] #mandatory
+# The modid of the mod
+modId="baritoe" #mandatory
```
Comment (ZeroMemes) [inline src/launch/resources/META-INF/mods.toml]: should be `baritone`

File: src/launch/resources/pack.mcmeta [hunk 14]
```diff
@@ -0,0 +1,6 @@
+{
+    "pack": {
+        "description": "baritoe",
```
Comment (ZeroMemes) [inline src/launch/resources/pack.mcmeta]: should be `baritone`

Comment (ZeroMemes) [reply src/launch/resources/pack.mcmeta]: also is this required by forge type poz?

Comment (babbaj) [reply src/launch/resources/pack.mcmeta]: not required but you get a warning screen

Comment (leijurv) [reply src/launch/resources/pack.mcmeta]: baritoe is correct

File: src/launch/resources/META-INF/mods.toml (hunk 13 above)
Comment (leijurv) [reply src/launch/resources/META-INF/mods.toml]: baritoe is correct

**Pull request merged**